pkg/api/bzz.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/ethersphere/bee/pkg/collection/entry"
+	"github.com/ethersphere/bee/pkg/encryption"
+	"github.com/ethersphere/bee/pkg/file"
+	"github.com/ethersphere/bee/pkg/file/joiner"
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/gorilla/mux"
+)
+
+const (
+	// ManifestContentType represents content type used for noting that specific
+	// file should be processed as manifest
+	ManifestContentType = "application/bzz-manifest+json"
+)
+
+func (s *server) bzzDownloadHandler(w http.ResponseWriter, r *http.Request) {
```

Comment:
This is a _very_ long function

Reply:
That it is. It was intentionally left like that, to have linear steps that are being performed. There is also a lot of boilerplate regarding HTTP API.
BTW, not the longest one, the [file.go](https://github.com/ethersphere/bee/blob/master/pkg/api/file.go) wins in that regard.

Reply:
well mistakes don't need to be repeated

Reply:
I would say this function is not specially long, and having a "long" function is not bad in any general way.  Also, we are generally trying to isolate the PRs to be small, and limit the changset, so I would not spend to much time here to refactor the `files.go` code. Especially, since we asked about this on mattermost and got the info that the code inside was not considered business logic and that it should be used in this way.
In order to refactor this function we should probably:
1. Extract specific error types
2. Remove logging and request rendering form the function and move it to calling functions, based on extracted error types
3. Maybe move this to a special upload/download/transfer component

So, anyways, my 2 c are that we should keep this PR minimal and focuse on the user-story, and open the separate PR to refactor the code, since this code will move stuff from `files.go` and possibily other api calls as well.